### ISSUE TITLE
Suggestion for more efficient window and application switching

### ISSUE THREAD

**Issue by adamsamec** [nvaccess/nvda#13913]
### Is your feature request related to a problem? Please describe.
When working with multiple applications and their many opened windows in Microsoft Windows using NVDA, I am missing an efficient way to quickly find and switch to the desired window.

1. In VoiceOver on macOS, there is the handy feature to invoke  a menu which offers a list of all running applications, allowing the user to switch to the selected application by pressing Enter on it, or expand a given application menu item with the Right arrow key to show the list of all opened windows only for that particular application. This menu can be invoked in VoiceOver via VO + F1 + F1 keyboard shortcut.
2. A second useful command in VoiceOver on macOS is a one which shows directly a menu of all windows only for the application in the foreground, allowing to switch to the selected window by pressing Enter. This menu can be invoked via VO + F2 + F2.

The benefits of these menus when compared with the classical way of window switching using Alt + Tab are mainly the following:

* Both of these menus support the filter menu items by typing feature. Therefore, say that I have a Google Chrome window in the foreground, with many other windows of Google Chrome open, and then I involke the abovementioned menu listing all the open windows of the app in the foreground, I can type "google search" and the menu will change to offer only the Google Chrome windows having the phrase "google search" in their title no matter whether in the begining of the title or the middle of it. The advantage of filtering the menu as you type is especially when you want to find a particular open browser window where you don’t recall the exact title of that browser window, but for instance, you just know the title contains the word „cooking", so after invoking the menu, you type just "cooking" and get the list of all the browser windows with the word "cooking" in it.
* The benefit of the firs command I mentioned above is that often you want to switch to some opened web page, and you know it is running in Microsoft Edge for instance, Therefore you can first find the Microsoft Edge submenu and then the specific window only within that application submenu. This way you don’t have to traverse all the opened window only to find that desired window, but, in other words, you can narrow down your search by certain application.
* The benefit of the second command I mentioned above, that is, the one for switching between only windows of the application in the foreground, is for instance when you want to copy something from one opened File explorer window to another, or when you want to get a list of all opened windows of a web browser and switch to one of them.

### Describe the solution you'd like
The same behavior as described in the previous section would be very welcome.

The keyboard shortcuts for invoking those menus should be as comfortable as possible. Perhaps NVDA + W could be a good candidate for the window list menu and NVDA + Shift + W for the application list menu.

### Describe alternatives you've considered
* In Windows 10 or 11, there are virtual desktops, which to some extent help the user to organize multiple applications and windows, but that’s far from the comfort of the menus I’ve described.
* Another way of achieving what I described above is to use the Task bar, Windows + T,, but that is lacking especially the following features:
  * It does not support first letter navigation, not saying the filter items by typing feature.
  * It shows not only the running applications, but also those pinned to Task bar, which get into way when you search for a particular running application.
  * The items in the Task bar are not ordered according to the time they were lastly used but according to the time those applications were started.
* There is also the Task View, invoked via Windows + Tab, but here, from the keyboard user perspective,  the windows are stil not organized by the applications they pertain to, neither it allows to used the first-letter navigation or filter windows by typing feature.

### Additional context
The feature request is inspired by VoiceOver on macOS.

**Comment by josephsl** [comment]
Hi, try Task View (Windows+Tab) and see if that makes a difference. NVDA+F11 is already taken by an add-on. Thanks.

**Comment by Brian1Gaff** [comment]
Have you encountered the  long running windows bug that in some instances 
alt tab cannot show you one of the opened windows? I see it a lot if I have 
a lot of windows open and find I have to close them one at a time until the 
one I need is visible again.
Its still around in windows 10, but not tried 11.
 Brian

-- 
***@***.***
Sent via blueyonder.(Virgin media)
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.
----- Original Message ----- 
From: "Adam" ***@***.***>
To: "nvaccess/nvda" ***@***.***>
Cc: "Subscribed" ***@***.***>
Sent: Sunday, July 17, 2022 12:06 AM
Subject: [nvaccess/nvda] Suggestion for more efficient window and 
application switching (Issue #13913)


### Is your feature request related to a problem? Please describe.
When working with multiple applications and their many opened windows in 
Microsoft Windows, I am missing an efficient way to quickly find and switch 
to the desired window.

1. As some of you might know, in VoiceOver on macOS, there is the handy 
possibility to invoke via VO + F1 + F1 a menu which offers a list of all 
running applications, allowing the user to switch to the selected 
application by pressing Enter on it, or expand a given application menu item 
with the Right arrow key to show the list of all opened windows only for 
that particular application.
2. A second useful command in VoiceOver on macOS, VO + F2 + F2, is a one 
which shows directly a menu of all windows for the application in the 
foreground, allowing to switch to the selected window by pressing Enter.

The benefits of these menus when compared with the classical way of window 
switching using Alt + Tab are mainly the following:

* Both of these menus support the filter menu items by typing feature. 
Therefore, say that I have a Google Chrome window in the foreground, with 
many other windows of Google Chrome open, and then I involke the 
abovementioned menu listing all the open windows of the app in the 
foreground, I can type "google search" and the menu will change to offer 
only the Google Chrome windows having the phrase "google search" in their 
title no matter whether in the begining of the title or the middle of it. 
The advantage of filtering the menu as you type is especially when you want 
to find an open website where you don’t recall the exact title of the 
browser window, but for instance, you just know the title contains the word 
„cooking".
* The benefit of the firs command I mentioned above is that often you want 
to switch to some opened web page, and you know it is running in Microsoft 
Edge for instance, Therefore you can first find the Microsoft Edge submenu 
and then the specific window only within that application submenu. This way 
you don’t have to traverse all the opened window only to find that desired 
window, but, in other words, you can narrow down your search by certain 
application.
* The benefit of the second command I mentioned above, that is, the one for 
switching between only windows of the application in the foreground is for 
instance when you want to copy something from one opened File explorer 
window to another, or when you want to get a list of all opened windows of a 
web browser and switch to one of them.

### Describe the solution you'd like
The same behavior as described in the previous section would be very 
welcome.

The keyboard shortcuts for invoking those menus should be as comfortable as 
possible. Perhaps NVDA + F11 could be a good candidate for the window list 
menu and NVDA + Shift + F11 for the application list menu.

### Describe alternatives you've considered
I know in Windows there are virtual desktops, which to some extent help the 
user to organize multiple applications and windows, but that’s far from the 
comfort of the menus I’ve described.

Another way of achieving what I described above is to use the Task bar, but 
that is lacking especially the following features:

* It does not support first letter navigation, not saying the filter items 
by typing feature.
* It shows not only the running applications, but also those pinned to Task 
bar, which get into way when you search for a particular running 
application.
* The items in the Task bar are not ordered according to the time they were 
lastly used but according to the time those applications were started.

### Additional context
The feature request is inspired by VoiceOver on macOS.

-- 
Reply to this email directly or view it on GitHub:
https://github.com/nvaccess/nvda/issues/13913

**Comment by adamsamec** [comment]
Hello @josephsl, I know about the Task View Windows feature invoked via Windows + Tab, but that seems not to offer any benefit over Alt + Tab, and is therefore still not as useful as the menus I described. I have updated the "Alternatives I've considred" section where I describe the lackings of the Task View feature.

I have also suggested a different keyboard shortcut for these proposed features: NvDA + W for the windows list menu and NvDA + Shift + W for the applications list menu.

**Comment by josephsl** [comment]
Hi,

In the past, Windows 10 did offer a timeline feature where you can view a history of activities you were performing on the computer. This is gone now (for the most part).

It might be possible to provide an "app windows manager" functionality for NVDA, preferably as an add-on (tied to an issue I'll discuss below). That way anyone requiring this feature can install the needed add-on without sacrificing the overall goal of a screen reader (first and foremost, NVDA is a screen reader, not a window management tool). We also need to remember that features like the one VoiceOver provide relies on what macOS gives, which is a different beast compared to what Windows says. Therefore, I advise contacting Microsoft by posting a Feedback Hub item so folks can upvote it (as window management functionality is really the responsibility of the window manager, in this case shell components from Windows).

As for add-ons and related issues: NVDA+W and NVDA+Shift+W are already taken by an add-on. As for this feature being part of an add-on, see if WinWizard is the one you might be interested in using.

Hope this helps. Thanks.

**Comment by josephsl** [comment]
Hi,

As a follow-up to my previous comment: I think what you might be really looking for is improved first-letter navigation and treating apps (especially Task View) as an enhanced list view. Note that Narrator also suffers from this issue, so it is something Microsoft should look at in the future.

Thanks.

**Comment by adamsamec** [comment]
As a first step, I think making the features I propose as a NVDA add-on is the way to go. Although, this does not necessarily need to be implemented as a NVDA add-on but could be a Windows utility, something like [Easy Window Switcher](https://neosmart.net/EasySwitch/#:~:text=Easy%20Window%20Switcher%20makes%20switching,only%20a%20alt%20%2B%20%60%20away!).

Expecting the proposed features to be implemented on the side of Microsoft Windows system would require to redesign the window switching philosophy from the grounds up, which I don't expect in the near future.

As for the keyboard shortcuts, with which add-on does the NVDA + W and NVDA + Shift + W shortcuts conflict? Perhaps the conflicting add-on could modify its keyboard shortcuts instead, to free them up for the features proposed in this issue.

I've checked what does the WinWizard NVDA add-on offers, but haven't found it as a satisfactory solution. Also, the proposal is not just to look for the first-letter navigation possiblity in the windows switcher list, but the support for the filter by typing feature is what would make a difference.

**Comment by josephsl** [comment]
Hi,

As for Task View and other changes from Microsoft: it requires refining the user interface library used in Task View and other parts of Windows shell so it can accept character input (last year, Windows Insiders hit a similar problem with quick link menu (Windows+X) where shortcut keys didn't work, which was fixed a few weeks later). At leat letting Microsoft know about ineffeciencies of Task View and other shell elements for keyboard users is, to me, the first step to go for his issue.
As for editing add-ons: the conflicting add-on is WeatherPlus. I think it would not be a helpful move to suggest that add-ons free up shortcuts for implementing what is being suggested in this issue - we can ask authors to consider keyboard command conflicts, but the decision is theirs (authors) to make, not us (really). But before we can discuss keyboard command conflicts, I think it would be best to at least suggest an add-on of some kind or enhancements to existing add-ons so the suggestion can come to life (at least in a form that can be tested).

Thanks.

**Comment by adamsamec** [comment]
I agree. Thanks, @josephsl . I will try to contact Microsoft and update this issue accordingly.

**Comment by seanbudd** [comment]
For the reasons @josephsl  mentions, I would suggest that Windows management is outside the scope of NVDA core.
I believe this needs to either be fixed on Windows end or as an NVDA Add-on.
